lntemp/node/watcher.go

```diff
@@ -0,0 +1,724 @@
+// the open channels.
+func (nw *nodeWatcher) updateNodeStateOpenChannel(op wire.OutPoint,
+	newChan *lnrpc.ChannelEdgeUpdate) {
+
+	// Load the old updates the node has heard so far.
+	updates := make([]*OpenChannelUpdate, 0)
+	result, ok := nw.state.openChans.Load(op)
+	if ok {
+		updates = result.([]*OpenChannelUpdate)
+	}
+
+	// Create a new update based on this newChan.
+	newUpdate := &OpenChannelUpdate{
+		AdvertisingNode: newChan.AdvertisingNode,
+		ConnectingNode:  newChan.ConnectingNode,
+		Timestamp:       time.Now(),
+	}
+
+	// Update the node's state.
+	updates = append(updates, newUpdate)
+	nw.state.openChans.Store(op, updates)
+
+	// For this new channel, if the number of edges seen is less
+	// than two, then the channel hasn't been fully announced yet.
+	if len(updates) < 2 {
+		return
+	}
+
+	// Otherwise, we'll notify all the registered watchers and
+	// remove the dispatched watchers.
+	watcherResult, loaded := nw.openChanWatchers.LoadAndDelete(op)
+	if !loaded {
+		return
+	}
+	events := watcherResult.([]chan struct{})
+	for _, eventChan := range events {
+		close(eventChan)
+	}
+}
+
+// updateNodeStatePolicy updates the internal state of the node regarding the
+// policy updates.
+func (nw *nodeWatcher) updateNodeStatePolicy(op wire.OutPoint,
+	newChan *lnrpc.ChannelEdgeUpdate) {
+
+	// Init an empty policy map and overwrite it if the channel point can
+	// be found in the node's policyUpdates.
+	policies := make(PolicyUpdate)
+	result, ok := nw.state.policyUpdates.Load(op)
+	if ok {
+		policies = result.(PolicyUpdate)
+	}
+
+	node := newChan.AdvertisingNode
+
+	// Append the policy to the slice and update the node's state.
+	newPolicy := PolicyUpdateInfo{
+		newChan.RoutingPolicy, newChan.ConnectingNode, time.Now(),
+	}
+	policies[node] = append(policies[node], &newPolicy)
+	nw.state.policyUpdates.Store(op, policies)
+}
+
+// handleOpenChannelWatchRequest processes a watch open channel request by
+// checking the number of the edges seen for a given channel point. If the
+// number is no less than 2 then the channel is considered open. Otherwise, we
+// will attempt to find it in its channel graph. If neither can be found, the
+// request is added to a watch request list than will be handled by
+// handleChannelEdgeUpdates.
+func (nw *nodeWatcher) handleOpenChannelWatchRequest(req *chanWatchRequest) {
+	targetChan := req.chanPoint
+
+	// If this is an open request, then it can be dispatched if the number
+	// of edges seen for the channel is at least two.
+	result, ok := nw.state.openChans.Load(targetChan)
+	if ok && len(result.([]*OpenChannelUpdate)) >= 2 {
+		close(req.eventChan)
+		return
+	}
+
+	// Before we add the channel to our set of open clients, we'll check to
+	// see if the channel is already in the channel graph of the target
+	// node. This lets us handle the case where a node has already seen a
+	// channel before a notification has been requested, causing us to miss
+	// it.
+	// TODO(yy): no longer needed?
```

Comment:
No longer needed as in "this can't happen anymore" or as in "this is checked somewhere else now"?